### PR TITLE
Plugins: Fix AudioControl and InFeedback processing for an extra cycle

### PR DESCRIPTION
**Pull request by vitreo12**
## Purpose and Motivation

I have noticed a weird behavior of the `AudioControl` and `InFeedback` UGens.

If you run this code, you'll notice that the last audio cycle will run twice, generating a "double" dropout:

```SuperCollider
(
s.waitForBoot({
	SynthDef(\test, {
		Out.ar(\out.ir(0), { SinOsc.ar(2) * EnvGen.ar(Env([1, 1], [0.2]), doneAction:2) })
	}).add;

	s.sync;

	~bus = Bus.audio(s);

	Pbind(
		\instrument, \test,
		\dur, 1,
		\out, ~bus.index,
		\addAction, \addToHead
	).play;

	~reader = { \in.ar }.play(args:[ \in, 'a' ++ ~bus.index]);

	{ In.ar(0) }.plot(0.5);
})
)
```

image

This happens because the `AudioControl` UGen (as well as the `InFeedback` one) only check if the bus has been touched in the current audio cycle. In my opinion, it should probably take into consideration if the bus has been touched in the previous audio cycle aswell, in order to avoid this "repetition" effect.

This PR fixes this issue. It basically allows `AudioControl` and `InFeedback` to work just like `In` when the `~reader` `Synth` is later in the chain, and just like `InFeedback` when the `~reader` `Synth` is before in the chain (reading the previous audio cycle).

`Pbind (\addToHead) -> ~reader:`
image (1)

`-> ~reader Pbind (\addToTail) ->:`
image (2)

## Types of changes

- Bug fix

## To-do list

- [X] Code is tested
- [X] All tests are passing
- [X] Updated documentation
- [x] This PR is ready for review

### CODE REVIEW

**Comment by vitreo12** [comment]
Sure, I just added some tests. Let me know if they are enough :)

**Comment by telephon** [comment]
I think these failing tests are just timeouts, @vitreo12. It would be good to get this working, perhaps you can see if there is something in your tests that causes it?

**Comment by vitreo12** [comment]
Hi @telephon! All the tests I have implemented complete on my local Linux build, I don't know why they would hang on MacOS. Could anyone with MacOS maybe try them out and report what is going wrong? 

The test class is called `TestIOUGens` and the functions are: `test_InFeedback_equals_In`, `test_AudioControl_equals_In`, `test_InFeedback_equals_delayed_In` and `test_AudioControl_equals_delayed_In`.

**Review by dyfer — CHANGES_REQUESTED** [review]
Hello @vitreo12 
I have a request regarding the tests: it seems that they use `Silence.ar`, which does not exist in the standard class library. Could you replace it with `Silent.ar`?

**Comment by dyfer** [comment]
@telephon @joshpar aside from the tests, do the changes in this PR look good to you?

**Comment by vitreo12** [comment]
> Hello @vitreo12 I have a request regarding the tests: it seems that they use `Silence.ar`, which does not exist in the standard class library. Could you replace it with `Silent.ar`?

Sure, just pushed in the changes.

File: testsuite/classlibrary/TestIOUGens.sc
Comment (telephon) [inline testsuite/classlibrary/TestIOUGens.sc:25]: The report should be clearer, something like: "InFeedback should be equal to In when no other signal has been written to its bus"

File: testsuite/classlibrary/TestIOUGens.sc
Comment (telephon) [inline testsuite/classlibrary/TestIOUGens.sc:59]: I'd suggest the message to say: "AudioControl should be equal to In for read after write".

File: testsuite/classlibrary/TestIOUGens.sc
Comment (telephon) [inline testsuite/classlibrary/TestIOUGens.sc:90]: suggestion: `test_AudioControl_equals_In_for_write_after_read`

**Review by telephon — COMMENTED** [review]
I see only some changes to be done in the naming of tests and messages. This is a very important fix, thanks a lot.

**Comment by smrg-lm** [comment]
**Please review if this change is necessary** or something else is failing because the given example is not good, it works as expected if the `test` ugen doesn't define a control input for the output bus, try replacing it by:

```
	SynthDef(\test, {
		Out.ar(0, SinOsc.ar(2) * EnvGen.ar(Env([1, 1], [0.2]), doneAction:2))
	}).add;
```

**Comment by vitreo12** [comment]
> **Please review if this change is necessary** or something else is failing because the given example is not good, it works as expected if the `test` ugen doesn't define a control input for the output bus, try replacing it by:
> 
> ```
> 	SynthDef(\test, {
> 		Out.ar(0, SinOsc.ar(2) * EnvGen.ar(Env([1, 1], [0.2]), doneAction:2))
> 	}).add;
> ```

That's because `AudioControl` or `InFeedback` are used for inner-Bus communication. With your code, you're just writing directly to the audio output.

**Comment by smrg-lm** [comment]
@vitreo12 I was wrong testing and the alarms sounded, apologies.

**Comment by vitreo12** [comment]
@telephon I have changed the names accordingly :)

**Review by dyfer — APPROVED** [review]

**Comment by dyfer** [comment]
Tests look good do me now.
@telephon feel free to merge if you're happy with the PR otherwise.

**Comment by smrg-lm** [comment]
Apologies for being annoying, I was worried because the use case is wrong, because the order of execution, and it makes the issue smell, but it smells good so far. I hope I'm not mistaken again: this change creates a gap in the audio in the following situation which is maybe better than what was happening before.

```
(
SynthDef('source', { arg out, freq=220;
	Out.ar(out, SinOsc.ar(freq));
}).add;

SynthDef('input', { arg a_in;
	Out.ar(0, a_in);
}).add;
)

b = Bus.audio();
x = Synth('source', [out: b, freq: 440]);
y = Synth('input', [a_in: b.asMap]);
z = Synth('source', [out: b]);

z.free;  // one block gap before x writes again.
```
After:

despues

Before:

antes

I'm not sure what could be the desired behavior in this particular case but the feedback was not intended to be used like this anyway.

**Comment by telephon** [comment]
That is a good point. It is actually a bit tricky what the best behaviour should be.

Possibiities:

1. all audio is read with time delay (this accumulates in multiple routing)
2. all audio is read with time delay only if there is at least one writer downbus, otherwise without (this may cause a discontinuity if the downbus writer stops)
3. the audio from upbus is read with no time delay and the audio from downbus is read with time delay
4. once the inFeedback has received its flag that there are writers downstream, it should continue to play delayed, irrespective of whether the downstream writers ended.

**Comment by joshpar** [comment]
As noted elsewhere, there likely isn't a way to satisfy ALL cases, but I do think this satisfies ones that make more sense than my initial implementation of this. I'm good approving it, but will wait until this weekend to merge in case @telephon or @dyfer feel like anything else can still be resolved.

**Review by joshpar — APPROVED** [review]

**Review by telephon — APPROVED** [review]

**Pull request merged**
